src/librustc_codegen_ssa/back/linker.rs

```diff
@@ -397,11 +397,6 @@ impl<'a> Linker for GccLinker<'a> {
     }
 
     fn export_symbols(&mut self, tmpdir: &Path, crate_type: CrateType) {
-        // Symbol visibility in object files typically takes care of this.
-        if crate_type == CrateType::Executable {
```

Comment:
As mentioned, avoiding this should be conditional on the target's `override_export_symbols` setting.